libvast/src/arrow_table_slice.cpp

```diff
@@ -1141,4 +1146,171 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
+    if (arr.IsNull(row)) {
+      if (!subnet_builder->AppendNull().ok())
+        die("append null value to subnet builder failed");
+    } else {
+      auto span = std::span<const uint8_t, 17>{arr.GetValue(row), 17};
+      if (!(subnet_builder->Append().ok()
+            && address_builder.Append(arr.GetValue(row)).ok()
+            && length_builder.Append(span[16]).ok()))
+        die("rebuilding of subnet struct array failed");
+    }
+  }
+  if (auto result = subnet_builder->Finish(); result.ok())
+    return *result;
+  die("failed to finish Arrow subnet column builder");
+}
+
+// TODO: this is copied from
+// experimental_table_slice_builder/enum_column_builder, can this live somewhere
+// shared?
+std::shared_ptr<arrow::Array> make_dict_array(const enumeration_type& et) {
+  arrow::StringBuilder dict_builder{};
+  for (const auto& f : et.fields())
+    if (!dict_builder.Append(std::string{f.name}).ok())
+      die("failed to build Arrow enum field array");
+  if (auto array = dict_builder.Finish(); array.ok()) {
+    return *array;
+  }
+  die("failed to finish Arrow enum field array");
+}
+
+/// Converts an array representing a single column from a previous arrow
+/// format to `arrow::experimental`.
+std::shared_ptr<arrow::Array>
+convert_column(const std::shared_ptr<arrow::Array>& arr, const type& t) {
+  auto f = detail::overload{
+    [&](const auto& t) -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(arr->type()->Equals(t.to_arrow_type()));
+      return arr;
+    },
+    [&](const record_type& rt) -> std::shared_ptr<arrow::Array> {
+      // this case handles VAST type `list<record>`
+      const auto& sa = static_pointer_cast<arrow::StructArray>(arr);
+      arrow::ArrayVector children{};
+      std::vector<std::string> field_names{};
+      children.reserve(rt.num_fields());
+      field_names.reserve(rt.num_fields());
+      int index = 0;
+      for (const auto& f : rt.fields()) {
+        const auto& arr = sa->field(index++);
+        children.push_back(convert_column(arr, f.type));
+        field_names.emplace_back(f.name);
+      }
+      auto res = arrow::StructArray::Make(children, field_names);
+      if (!res.ok())
+        die("unable to construct nested struct array");
+      return *res;
+    },
+    [&](const pattern_type&) -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<pattern_type::array_type>(
+        pattern_type::to_arrow_type(), arr);
+    },
+    [&](const address_type&) -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<address_type::array_type>(
+        address_type::to_arrow_type(), arr);
+    },
+    [&](const subnet_type&) -> std::shared_ptr<arrow::Array> {
+      auto ba = static_pointer_cast<arrow::FixedSizeBinaryArray>(arr);
+      VAST_ASSERT(ba->byte_width() == 17);
+      return convert_subnet_array(*ba);
+    },
+    [&](const enumeration_type& et) -> std::shared_ptr<arrow::Array> {
+      auto dict = make_dict_array(et);
```

Comment:
You should not need this, as this is done in the constructor of the builder already. If you do `et.to_arrow_builder()` you get a custom builder that inherits from the `arrow::DictionaryBuilder` with all the fields pre-configured.

Reply:
If the slice append method just calls Append multiple times under the hood, we can also just call our `enumeration_type::arrow_builder::Append` function directly with `detail::narrow_cast<enumeration>(value)` in a loop. Could probably throw in a `Reserve` once in the beginning for good measure to avoid re-allocations. Open to pairing on this quickly if this is unclear.

Reply:
The current implementation doesn't have to loop at all, because we directly construct the `DictionaryArray` from its components. Since enum is one of the columns that actually occur with some frequency in the current schemas, I'd like to avoid all potential costly rebuilding. I prefer explicitely building the dictionary array to keep that performance over going through the builder.

I came up with
```c++
      auto b = et.make_arrow_builder(arrow::default_memory_pool());
      auto empty_result = b->Finish();
      if (!empty_result.ok()) {
        die("unable to finish empty enum dictionary builder");
      }
      const auto& dict
        = static_cast<const arrow::DictionaryArray&>(
            *static_cast<const arrow::ExtensionArray&>(**empty_result).storage())
            .dictionary();
```

to leverage the current builder implementation or dictionary creation while avoiding the full rebuild. This is a bit ugly, but maybe we can massage that to make it better?

Reply:
I came up with this diff. Still kind of unergonomic, but I think this is at least somewhat better.

```diff
diff --git a/libvast/include/vast/type.hpp b/libvast/include/vast/type.hpp
index 90b095fe0..151354ea8 100644
--- a/libvast/include/vast/type.hpp
+++ b/libvast/include/vast/type.hpp
@@ -937,6 +937,11 @@ public:
 
     [[nodiscard]] std::shared_ptr<arrow::DictionaryArray> storage() const;
 
+    /// Create an array from a prepopulated indices Array.
+    static arrow::Result<std::shared_ptr<enumeration_type::array_type>>
+    make(const std::shared_ptr<enumeration_type::arrow_type>& type,
+         const std::shared_ptr<arrow::UInt8Array>& indices);
+
   private:
     using arrow::ExtensionArray::storage;
   };
@@ -997,6 +1002,7 @@ public:
 struct enumeration_type::arrow_type final : arrow::ExtensionType {
   friend class type;
   friend struct enumeration_type::builder_type;
+  friend struct enumeration_type::array_type;
 
   /// A unique identifier for this extension type.
   static constexpr auto name = "vast.enumeration";
diff --git a/libvast/src/type.cpp b/libvast/src/type.cpp
index f7db35af1..6d620b8c5 100644
--- a/libvast/src/type.cpp
+++ b/libvast/src/type.cpp
@@ -2135,6 +2135,24 @@ void enumeration_type::arrow_type::register_extension() noexcept {
   VAST_ASSERT(status.ok());
 }
 
+arrow::Result<std::shared_ptr<enumeration_type::array_type>>
+enumeration_type::array_type::make(
+  const std::shared_ptr<enumeration_type::arrow_type>& type,
+  const std::shared_ptr<arrow::UInt8Array>& indices) {
+  auto dict_builder
+    = string_type::make_arrow_builder(arrow::default_memory_pool());
+  for (const auto& [canonical, internal] : type->vast_type_.fields()) {
+    const auto append_status = dict_builder->Append(
+      arrow::util::string_view{canonical.data(), canonical.size()});
+    VAST_ASSERT(append_status.ok(), append_status.ToString().c_str());
+  }
+  ARROW_ASSIGN_OR_RAISE(auto dict, dict_builder->Finish());
+  ARROW_ASSIGN_OR_RAISE(auto storage,
+                        arrow::DictionaryArray::FromArrays(
+                          string_type::to_arrow_type(), indices, dict));
+  return std::make_shared<enumeration_type::array_type>(type, storage);
+}
+
 enumeration_type::builder_type::builder_type(std::shared_ptr<arrow_type> type,
                                              arrow::MemoryPool* pool)
   : arrow::StringDictionaryBuilder(string_type::to_arrow_type(), pool),
```

Reply:
this mostly works, and it's better (as in, more centralized).
Adopted, and thanks!